realm/src/main/java/io/realm/RealmObject.java

```diff
@@ -124,4 +135,129 @@ protected static Realm getRealm(RealmObject obj) {
     protected static Row getRow(RealmObject obj) {
         return obj.row;
     }
+
+    /**
+     * Set the Future instance returned by the worker thread, we need this instance
+     * to force {@link #load()} an async query, we use it to determine if the current
+     * RealmResults is a sync or async one.
+     *
+     * @param pendingQuery pending query.
+     */
+    void setPendingQuery(Future<Long> pendingQuery) {
+        this.pendingQuery = pendingQuery;
+        if (isLoaded()) {
+            // the query completed before RealmQuery
+            // had a chance to call setPendingQuery to register the pendingQuery (used btw
+            // to determine isLoaded behaviour)
+            onCompleted();
+
+        } // else, it will be handled by the Realm#handler
+    }
+
+    /**
+     * Determine if the current RealmObject is obtained synchronously or asynchronously (from
+     * a worker thread). Synchronous RealmObjects are by definition blocking hence this method
+     * will always return {@code true} for them.
+     *
+     * @return {@code true} if the query has completed & the data is available {@code false} if the
+     * query is in progress.
+     */
+    public boolean isLoaded() {
+        realm.checkIfValid();
```

Comment:
This will crash for stand-alone objects. For those this method should just return `false` IMO